### ISSUE TITLE
IADB_MNIST.py example - possible bug in model

### ISSUE THREAD

**Issue by cpldcpu** [tchambon/posts#1]
Hi,

first, thank you a lot for the great alpha deblending paper and also the minimal MNIST example. It's really great to have an introductionary resource on diffusion models that don't start with trying to explain the maths behind the DDPM/DDIM noise schedulers.

I have been playing with the MNIST sample and noticed a problem when adopting it to other datasets: The currently used U-Net reduces to a minimum dimensionality of 64x4x4, but there does not seem to be any layer that "communicates" between the 4x4 regions? 

I introduced FC layers in the middle, which seems to bring some improvement.

```python
        # fc layer in class definition
        self.fc1 = torch.nn.Linear(4*4*64, 4*4*64)
        self.fc2 = torch.nn.Linear(4*4*64, 4*4*64)
...
        # forward pass
        m_fc1 = self.relu(self.fc1(b3_c4.view(-1, 4*4*64)))
        m_fc2 = self.relu(self.fc2(m_fc1))
        m_fc2 = m_fc2.view(-1, 64, 4, 4)
        u2_c1 = self.relu(self.block3_up1(m_fc2))# + b3_c3

```
I noticed the issue with a synthetic dataset showing dice. With the orignal code, their shape is not retained, while the version with the fc layers looks correct.

Orignal:
![generated_dice_31_nofc](https://github.com/tchambon/posts/assets/4086406/a98d7dae-42de-4737-b2a8-4331a2c3a4a3)

Modified:
![generated_dice_31_with_fc](https://github.com/tchambon/posts/assets/4086406/75d846c3-9ba6-4664-9fca-fba19cc96539)

The differfence is not that notable with the MNIST dataset, but that is probably because there are many training images that barely look like numbers...

Original:
![generated_mnist_127_original](https://github.com/tchambon/posts/assets/4086406/821d6bc0-0417-483e-9c47-e237ea03d65e)

Modified:
![generated_mnist_127_fixed](https://github.com/tchambon/posts/assets/4086406/9f7907e5-4c92-412b-b698-c65cca57be12)

**Comment by tchambon** [comment]
Hi,

Thank you for reaching out.

The architecture presented here is intentionally basic, serving primarily as a means to demonstrate a straightforward diffusion tutorial. It's true that there are numerous ways to enhance it, such as adding more layers at the Unet bottleneck, which you've mentioned.

We reserve the more sophisticated architectures, like the Unet2D from the diffusers package, for our main repository at https://github.com/tchambon/IADB. 
These are not included in this simple tutorial because our aim here is to focus on the concept of IADB loss rather than architectural complexity, which is sufficient for the purposes of teaching with a dataset as simple as MNIST.

I hope this helps clarify the purpose of the architecture used in this tutorial.
